### PR TITLE
modules: Clean up modules' errors

### PR DESCRIPTION
**Pull request by yito88**
Closes: #1333 
based on #1283

## Description
- Add ~~`Ics02Error` and~~ `Ics03Error` to `Ics04Error`
- Give the port ID and the channel ID to `MissingNext*Seq` error
- Remove redundant errors

______

For contributor use:

- [x] Added a changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [x] If applicable: Unit tests written, added test to CI.
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Updated relevant documentation (`docs/`) and code comments.
- [x] Re-reviewed `Files changed` in the Github PR explorer.

### CODE REVIEW

File: modules/src/ics04_channel/error.rs [hunk 1]
```diff
@@ -12,6 +13,14 @@ use tendermint_proto::Error as TendermintError;
 define_error! {
     #[derive(Debug, PartialEq, Eq)]
     Error {
+        Ics02Client
+            [ client_error::Error ]
+            | _ | { "ics02 client error" },
+
```
Comment (hu55a1n1) [inline modules/src/ics04_channel/error.rs:19]:
```suggestion
```
I would suggest removing this variant from here and reusing `Ics03Connection::Ics02Client`.

File: modules/src/mock/context.rs [hunk 2]
```diff
@@ -446,66 +450,116 @@ impl PortReader for MockContext {
...
     }
 
     fn client_consensus_state(
         &self,
         client_id: &ClientId,
         height: Height,
-    ) -> Option<AnyConsensusState> {
-        ClientReader::consensus_state(self, client_id, height)
+    ) -> Result<AnyConsensusState, Ics04Error> {
+        ClientReader::consensus_state(self, client_id, height).map_err(Ics04Error::ics02_client)
```
Comment (hu55a1n1) [inline modules/src/mock/context.rs:486]:
```suggestion
        ClientReader::consensus_state(self, client_id, height)
            .map_err(|e| Ics04Error::ics03_connection(Ics03Error::ics02_client(e)))
```

File: modules/src/mock/context.rs [hunk 3]
```diff
@@ -446,66 +450,116 @@ impl PortReader for MockContext {
...
+        match self.connection_channels.get(cid) {
+            Some(pcid) => Ok(pcid.clone()),
+            None => Err(Ics04Error::missing_channel()),
+        }
     }
 
-    fn client_state(&self, client_id: &ClientId) -> Option<AnyClientState> {
-        ClientReader::client_state(self, client_id)
+    fn client_state(&self, client_id: &ClientId) -> Result<AnyClientState, Ics04Error> {
+        ClientReader::client_state(self, client_id).map_err(Ics04Error::ics02_client)
```
Comment (hu55a1n1) [inline modules/src/mock/context.rs:478]:
```suggestion
        ClientReader::client_state(self, client_id)
            .map_err(|e| Ics04Error::ics03_connection(Ics03Error::ics02_client(e)))
```

**Review by hu55a1n1 — CHANGES_REQUESTED** [review]

**Review by hu55a1n1 — CHANGES_REQUESTED** [review]
Thanks @yito88! Just one more thing - compilation is currently failing for tests. Here's a fix ->
```patch
diff --git a/modules/src/ics04_channel/handler/chan_open_try.rs b/modules/src/ics04_channel/handler/chan_open_try.rs
index dd71eb65..5191e0ef 100644
--- a/modules/src/ics04_channel/handler/chan_open_try.rs
+++ b/modules/src/ics04_channel/handler/chan_open_try.rs
@@ -370,12 +370,16 @@ mod tests {
                 msg: ChannelMsg::ChannelOpenTry(msg.clone()),
                 want_pass: false,
                 match_error: Box::new(|e| match e {
-                    error::ErrorDetail::Ics02Client(e) => {
+                    error::ErrorDetail::Ics03Connection(e) => {
                         assert_eq!(
                             e.source,
-                            ics02_error::ErrorDetail::ClientNotFound(
-                                ics02_error::ClientNotFoundSubdetail {
-                                    client_id: ClientId::new(ClientType::Mock, 45).unwrap()
+                            ics03_error::ErrorDetail::Ics02Client(
+                                ics03_error::Ics02ClientSubdetail {
+                                    source: ics02_error::ErrorDetail::ClientNotFound(
+                                        ics02_error::ClientNotFoundSubdetail {
+                                            client_id: ClientId::new(ClientType::Mock, 45).unwrap()
+                                        }
+                                    )
                                 }
                             )
                         );
```
Once #1283 is merged, it would be great if you could rebase on/merge with master.

**Comment by yito88** [comment]
Thank you for reviewing and fixing!
I've rebased on master.

**Review by hu55a1n1 — APPROVED** [review]

**Review by adizere — APPROVED** [review]
Nice idea to consolidate the errors like this. Great work!

**Pull request merged**